FEATURE: Synchronous and asynchronous process manager

robertlemke opened the pull request:
This change introduce a basic process manager which can keep
a state containing further information about the process at hand.

In order to make the process manager more convenient, this change
also  introduces a new interface which allows for custom handling
of event listener method invocations.

Those event listeners implementing this interface gain control
over the actual event listener method call and are able to
take additional steps to prepare for handling the event.

bwaidelich commented on Classes/Event/EventPublisher.php:
```diff
@@ -104,7 +105,11 @@ public function publishMany(string $streamName, array $events, int $expectedVers
             $event = $this->propertyMapper->convert($rawEvent->getPayload(), $eventClassName, $configuration);
             foreach ($this->eventListenerLocator->getSynchronousListeners($rawEvent->getType()) as $listener) {
-                call_user_func($listener, $event, $rawEvent);
+                if ($listener[0] instanceof EventListenerInvocationInterface) {
```
`$listener` is an instance of [callable](http://php.net/manual/en/language.types.callable.php). PHPs implementation of that is a bit ...weird.. but it could be a closure or an array ['<className>', '<methodName>'] for example

bwaidelich commented on Classes/EventListener/EventListenerInvocationInterface.php:
```diff
@@ -0,0 +1,33 @@
+     * @param RawEvent $rawEvent The raw event to be dispatched
+     * @return void
+     */
+    public function invokeEventListenerMethod(string $eventListenerMethodName, EventInterface $event, RawEvent $rawEvent);
```
When we talked about it we thought about a static constructor method that would be used in place of `ObjectManager::get()` if the listener class implements a specific interface, sth. like:
```php
public static function createFromEvent(EventInterface $event, RawEvent $rawEvent)
{
}
```

wouldn't that work for your case?

bwaidelich commented on Classes/ProcessManager/AbstractProcessManager.php:
```diff
@@ -0,0 +1,87 @@
+        $this->state = $this->stateRepository->get($stateIdentifier, get_class($this));
+        if ($this->state === null) {
+            $this->state = new State($stateIdentifier, get_class($this));
+        }
```
..the code above would then go into the constructor (or some `initializeObject()` method) and we could keep the existing `when*()` implementation

robertlemke replied:
As far as I can see, that wouldn't help in my case because I need to prepare the process state individually for each `when*()` call, not only once when the process manager is instantiated. And furthermore, my process manager currently is a singleton (because I can! ;-) ).

bwaidelich replied:
Mh, in that case I'd still prefer a generic `preXYZ` method, leaving the listener invocation to the framework.
Either by some interface or maybe even using some annotation on the method that expects `EventInterface` and (optionally) `RawEvent`

albe replied:
And I'd rather have such specific requirements delt with by the specific EventListener itself, than introducing new Interfaces and generic hook callbacks all around the place. It just adds accidential complexity in the core infrastructure for specific use cases.

Anyone has any real concerns/issues about or arguments against going the `EventListenerInterface` = `handle(EventInterface)` route and keeping the actual handler method invocation & lifecycle callbacks to the conrete listener implementations?

albe replied:
along those lines:

```
interface EventListenerInterface {
    public function handle(EventInterface $event, ...);
}
```

```
abstract class AbstractWhenMethodConventionBasedEventListener implements EventListenerInterface {

    public function handle(EventInterface $event, ...) {
        $eventClassName = get_class($event);
        $handlerMethod = 'when' . (new \ReflectionClass($eventClassName))->getShortName();
        if (is_callable($this, $handlerMethod)) {
            return $this->$handlerMethod($event, ...);
        }
        return false;
    }

}
```

```
class EventPublisher ... {

...
   foreach ($this->eventListenerLocator->getSynchronousListeners($rawEvent->getType()) as $listener) {
       call_user_func([$listener, 'handle'], $event, $rawEvent);
   }

}
```

Then you can simply implement the process manager or whatever other listener that needs pre-/post- handle hooks like this:

```
class ProcessManager extends AbstractWhenMethodConventionBasedEventListener {

   public function handle(EventInterface $event, ...) {
      if (is_callable([$this, 'beforeHandle']) {
         $this->beforeHandle($event);
      }
      return parent::handle($event, ...);
   }

}
```

Pro:
 - we get language-level functionality contracts for EventListeners
 - we shift some responsibility off the core infastructure (event storage, event publishing)
 - EventListeners can be more freely implemented to needs
 - No additional interfaces required for the infrastructure to support use cases like this one

Con:
 - we add an additional abstract class for EventListeners (could also be solved by a trait in this case, ie `use WhenMethodBasedEventHandling;`)
 - ...?

bwaidelich replied:
Thanks for clarifying, but I tend to disagree:
I think concise interfaces are *much* better than a deep inheritance hierarchy as it increases coupling and makes extension points less explicit.
I do get your point though and I agree that the currently hard-coded "when*" logic could be extended via some strategy pattern. But IMO it should still be the `EventListenerLocator`s responsibility to resolve that to a callable that we just invoke.

Just to be clear: I don't think it's a problem to create common base classes implementing those interfaces if it makes sense. But the logic should be based on the contracts and not on (or in) the abstract class IMO

robertlemke commented:
I changed the interface and pre-invocation mechanism so that the `EventPubslisher` is still in charge of calling the actual event listener method. I think that's a good compromise now and supports what we need in the process manager.

In general, I would like to merge this pull request as soon as we can. I know that the API won't be final, in fact I think that the whole state and configuration part will still change again, once we have correlation support. But I have more features coming up based on this, so for practical reasons I'd like to have this commit in master so in doesn't get messy …

bwaidelich commented on Classes/Event/EventPublisher.php:
```diff
@@ -106,6 +107,9 @@ public function publishMany(string $streamName, array $events, int $expectedVers
             $eventClassName = $this->eventTypeResolver->getEventClassNameByType($rawEvent->getType());
             $event = $this->propertyMapper->convert($rawEvent->getPayload(), $eventClassName, $configuration);
             foreach ($this->eventListenerLocator->getSynchronousListeners($rawEvent->getType()) as $listener) {
+                if ($listener[0] instanceof ActsBeforeInvokingEventListenerMethodsInterface) {
```
Just in case you missed it before: `$listener` is not necessarily an array in the form `[<object>, '<methodName>']` but could be any `\callable`. (just mentioning it for the record, I'm not sure how we could solve it)

bwaidelich commented on Classes/EventListener/ActsBeforeInvokingEventListenerMethodsInterface.php:
```diff
@@ -0,0 +1,33 @@
+     * @param RawEvent $rawEvent The raw event to be dispatched
+     * @return void
+     */
+    public function beforeInvokingEventListenerMethod(string $eventListenerMethodName, EventInterface $event, RawEvent $rawEvent);
```
Do we really need the `eventListenerMethodName`? IMO all relevant information should be in the `RawEvent` (see below)

bwaidelich replied:
You don't need to know the convention. But in the specific listener you know which method will handle which event. In any case that shouldn't matter though. Link to the event class name or type. Those stay the same (only that the type can in theory be mapped to a different class, so both can make sense)

bwaidelich commented on Classes/ProcessManager/AbstractProcessManager.php:
```diff
@@ -0,0 +1,85 @@
+    public function beforeInvokingEventListenerMethod(string $eventListenerMethodName, EventInterface $event, RawEvent $rawEvent)
+    {
+        $configuration = $this->getProcessConfiguration();
+        $eventName = substr($eventListenerMethodName, 4);
```
In other places we refer to the event *type* and this is stored in the `$rawEvent`.
The listener method name is determined by the short class name of the `EventInterface`:
```php
'when' . (new \ReflectionClass($eventClassName))->getShortName();
```

robertlemke replied:
yes. So, should we duplicate this logic or transfer it via the `beforeInvoking…` method calls?

bwaidelich replied:
I would say: Use `$rawEvent->getType()` as configuration identifier, no need to repeat anything then

bwaidelich commented on Classes/ProcessManager/ProcessManagerException.php:
```diff
@@ -0,0 +1,20 @@
+use Neos\Cqrs\Exception;
+
+/**
+ * Exception
```
That's rather sparse ;)

robertlemke replied:
I could repeat the class name in that doc comment – do you mean that?

robertlemke commented:
This is now a working solution for synchronous and asynchronous process managers.
From the last commit:

> This change introduces a new catch up command for
asynchronous event listeners (including process
managers) and a row of related features and changes.

> The new CLI command is located in the EventStore
command controller (`eventstore:catchup`) and
basically works like the `projection:catchup`
command except that you can't specify a specific
event listener.

> The log about applied events is stored in a new
generic "applied events log repository". This
repository replaces the specialized ProjectionState
repository used by the `AbstractDoctrineProjector`.
It is not mandatory to use this new repository,
but very convenient if you don't want to implement
your own storage.

> Finally, there are a few clean ups and new methods
in the Event Publisher which hopefully make it a
bit clearer how the respective methods work.

I'm aware of that we may or may not find a better place for the new `catchup` command than `eventstore:catchup`, but let's start with this version and then improve from there.

bwaidelich commented on Classes/Event/EventPublisher.php:
```diff
@@ -105,7 +106,10 @@ public function publishMany(string $streamName, array $events, int $expectedVers
-            foreach ($this->eventListenerLocator->getSynchronousListeners($rawEvent->getType()) as $listener) {
+            foreach ($this->eventListenerLocator->getSynchronousListenersByEventType($rawEvent->getType()) as $listener) {
+                if ($listener[0] instanceof ActsBeforeInvokingEventListenerMethodsInterface) {
+                    $listener[0]->beforeInvokingEventListenerMethod($listener[1], $event, $rawEvent);
```
`$listener` is still not an array (I keep repeating myself in case you didn't see it - as you didn't respond yet).
With the current implementation it *is* an array, but the signature says `\callable` and the idea was that you could have static or closure listeners.. It might be a case of YAGNI, but maybe we should at least add a comment here?

robertlemke replied:
yeah, I read it … I just wondered: you (we all) are using it as an array elsewhere already, so that would have to be changed, too, right? So should we check if it's an array and if it's not, ignore it in the relevant parts? 

Actually, I think it's YAGNI if nobody needed it **yet**. If we don't have a case for it right now, I'd rather remove it and add support for closures later again. As far as I can see it doesn't have an impact on userland anyway, right?

bwaidelich replied:
> you (we all) are using it as an array elsewhere already

No, we only use it as `call_user_func($listener, $event, $rawEvent);` which supports all `\callable`s

> So should we check if it's an array and if it's not, ignore it in the relevant parts?

Even if it's an array it could be in the form `['<ClassName>', '::<MethodName>']`..

> Actually, I think it's YAGNI

Probably. The thinking behind it was: There should be only one place where a listener is resolved. and the common "interface" for a listener would be something that you can just invoke. Maybe we can find a better way than `\callable`

> As far as I can see it doesn't have an impact on userland anyway, right?

Right, maybe add a comment anyways in case we change this at some point?

robertlemke commented:
@bwaidelich I added some safe guards for the potential Closure event listeners.

I think I'm fine with this change now ;-)

bwaidelich commented on Classes/Command/EventStoreCommandController.php:
```diff
@@ -113,4 +120,17 @@ public function dropSchemaCommand()
+     *
+     * @return void
+     */
+    public function catchUpCommand()
```
In general I hope that we can combine the concepts of projections & listeners where they overlap.
We might even get rid of the `ProjectionManager` and just have a slim logic in the `ProjectionCommandController` that filters the listeners that are projections and then uses the new `EventListenerManager` to catchup.
Then the only difference between a projection and an event listener is that the former is "replayable" (which is a special form of catchup really)

bwaidelich replied:
In general I hope that we can combine the concepts of projections & listeners where they overlap.
We might even get rid of the `ProjectionManager` and just have a slim logic in the `ProjectionCommandController` that filters the listeners that are projections and then uses the new `EventListenerManager` to catchup.
Then the only difference between a projection and an event listener is that the former is "replayable" (which is a special form of catchup really)

robertlemke replied:
Yeah, I fully agree - there should be just one command for catch up and watch which applies to all relevant event listeners.

I already started streamlining that - maybe put that into a new pull request?

robertlemke commented on Classes/EventListener/EventListenerManager.php:
```diff
@@ -140,6 +191,51 @@ public function getEventTypesByListenerClassName(string $listenerClassName): arr
     }
 
     /**
+     * @return int
```
yes! Let's improve it in a new PR - I'll prepare something.

robertlemke commented:
@bwaidelich okay, here we have an initial version of the `event:catchup` and `event:watch` commands, similar to the commands you introduced for projections. I have lots of ideas how to improve this, but well, let's put a lid on this an start with a fresh PR.

Feel free to merge if all looks fine, I won't push anything new today anymore.

robertlemke commented:
Good morning! 🌞 

I put some more love into the `catchup` and `watch` commands - for the time being only for the generic `event:*` commands because I'd argue that we can drop the `projection:catchup` and `projection:watch` commands at some point.

I introduced a new `--verbose` flag and changed the standard output as follows:

Without any further options, you'll see a condensed continuous progress, where "." marks an idle catch up cycle and "*" marks an event being applied:

<img width="959" alt="screen shot 2016-11-18 at 08 56 10" src="https://cloud.githubusercontent.com/assets/95582/20423084/977ee834-ad6e-11e6-98a3-12c8482ce0f0.png">

With the `--verbose` option you'll see, well, some more verbose output:

![image](https://cloud.githubusercontent.com/assets/95582/20423067/78607fa8-ad6e-11e6-9b25-49c494b3b567.png)

(ignore the `--exit-after` option in the screenshot - I removed support for that again before I pushed the last commit)

albe commented on Classes/EventListener/EventListenerManager.php:
```diff
@@ -24,7 +28,7 @@
  * @Flow\Scope("singleton")
  */
-class EventListenerLocator
+class EventListenerManager
```
I actually think `Locator` was a more expressive name to what this is/was about. `Manager` is rather arbitrary and the need to change probably came from adding the responsibility to `catch up`. But this breaks SRP, so maybe we should just move that to a more concrete service, like `EventListenerConsistencyManager`?

robertlemke replied:
I agree, the naming isn't optimal and also agree that it breaks the single-responsibility-principle. 

I'm not sure about `EventListenerConsistencyManager` though. How about moving `catchUp` to the `EventPublisher`. That's much closer to the topic.

albe commented on Classes/EventListener/AppliedEventsLog.php:
```diff
@@ -15,17 +15,17 @@
  * @Flow\Entity
  */
-class ProjectionState
+class AppliedEventsLog
```
I like that new name, much better!

albe commented on Classes/ProcessManager/State/State.php:
```diff
@@ -0,0 +1,215 @@
+ *
+ * @Flow\Entity
+ */
+class State
```
Maybe change to `ProcessState`? I actually think it's rather a single Process' State, not of the whole ProcessManager, no?

robertlemke replied:
Yes, I actually named it `ProcessState` at some point, don't know why it got lost.

albe commented:
Left a few more comments, but looks good overall.

robertlemke commented:
@albe thanks for your feedback! I implemented both suggestions in the last two commits.

albe commented:
Thanks! Makes sense somehow to put `catchUp` into the publisher, at least for local listeners. :+1: from my side

robertlemke commented:
All right, thanks y'all for your reviews! I merged this pile now, just let me know what else needs to be addressed in future pull requests.
